File: src/client/src/DataTransfer.ts
export type Event =
  { type: 'hidden' } & HiddenEvent
  | { type: 'released' } & ReleasedEvent
export type HiddenEvent = {
  title: string
  reservationStartTime: string
}
export type ReleasedEvent = {
  title: string
  infoText: string
  slots: Slot[]
}

export type Slot = {
  startTime: string
  duration: string | null
  type: SlotType
}

export type SlotType = {
  type: 'free'
  url: string
  closingDate: string | null
  maxQuantityPerBooking: number | null
  remainingCapacity: number | null
} | {
  type: 'closed'
} | {
  type: 'taken'
}

export type BookingResult = {
  slotType: SlotType
  mailSendError: boolean
}

export type BookingError =
  | { error: 'event-not-found' }
  | { error: 'event-not-released' }
  | { error: 'slot-not-found' }
  | { error: 'slot-not-free', slotType: SlotType }
  | { error: 'invalid-subscription-quantity' }
  | { error: 'invalid-subscriber-name' }
  | { error: 'invalid-mail-address' }
  | { error: 'invalid-phone-number' }
  | { error: 'max-quantity-per-booking-exceeded' }
  | { error: 'capacity-exceeded', slotType: SlotType }

File: src/server/DataTransfer.fs
namespace HTLVB.RegistrationForm.Server.DataTransfer

open System
open System.Text.Json.Serialization

[<AbstractClass>]
[<JsonPolymorphic(TypeDiscriminatorPropertyName = "type")>]
[<JsonDerivedType(typeof<SlotTypeFree>, "free")>]
[<JsonDerivedType(typeof<SlotTypeTakenWithRequestPossible>, "takenWithRequestPossible")>]
[<JsonDerivedType(typeof<SlotTypeTaken>, "taken")>]
[<JsonDerivedType(typeof<SlotTypeClosed>, "closed")>]
type SlotType() = class end
and SlotTypeFree(url: string, closingDate: Nullable<DateTime>, maxQuantityPerBooking: Nullable<int>, remainingCapacity: Nullable<int>) =
    inherit SlotType()
    member _.Url = url
    member _.ClosingDate = closingDate
    member _.MaxQuantityPerBooking = maxQuantityPerBooking
    member _.RemainingCapacity = remainingCapacity
and SlotTypeClosed() =
    inherit SlotType()
and SlotTypeTakenWithRequestPossible(url: string, closingDate: Nullable<DateTime>, maxQuantityPerBooking: Nullable<int>) =
    inherit SlotType()
    member _.Url = url
    member _.ClosingDate = closingDate
    member _.MaxQuantityPerBooking = maxQuantityPerBooking
and SlotTypeTaken() =
    inherit SlotType()

type Slot = {
    StartTime: DateTime
    Duration: Nullable<TimeSpan>
    Type: SlotType
}

[<AbstractClass>]
[<JsonPolymorphic(TypeDiscriminatorPropertyName = "type")>]
[<JsonDerivedType(typeof<HiddenEvent>, "hidden")>]
[<JsonDerivedType(typeof<ReleasedEvent>, "released")>]
type Event() = class end
and HiddenEvent(title: string, reservationStartTime: DateTime) =
    inherit Event()
    member _.Title = title
    member _.ReservationStartTime = reservationStartTime
and ReleasedEvent(title: string, infoText: string, slots: Slot[]) =
    inherit Event()
    member _.Title = title
    member _.InfoText = infoText
    member _.Slots = slots

type Subscriber = {
    Quantity: int
    Name: string
    MailAddress: string
    PhoneNumber: string
}

type BookingResult = {
    SlotType: SlotType
    MailSendError: bool
}
type RequestBookingResult = {
    MailSendError: bool
}